File: seminar-12/C/python/C.py
x = input()
y = input()

a = [[0] * (len(y) + 1) for _ in range(len(x) + 1)]

for i in range(len(x)):
    for j in range(len(y)):
        if x[i] == y[j]:
            a[i + 1][j + 1] = a[i][j] + 1
        else:
            a[i + 1][j + 1] = max(a[i][j + 1], a[i + 1][j])

print(a[len(x)][len(y)])

File: seminar-12/C/cpp/C.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string x, y;
    cin >> x >> y;

    vector<vector<int>> dp(x.size() + 1, vector<int>(y.size() + 1, 0));

    for (int i = 0; i < (int) x.size(); i++) {
        for (int j = 0; j < (int) y.size(); j++) {
            if (x[i] == y[j]) {
                dp[i + 1][j + 1] = dp[i][j] + 1;
            } else {
                dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
    }

    cout << dp[x.size()][y.size()] << '\n';
    return 0;
}
